text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,179 @@
+
+### pack
+
+Certain `pack` CLI arguments would take URIs supported instead of the current inconsistent format.
+
+Some examples:
+
+```shell script
+pack build my-app \
+  --buildpack urn:cnb:builder:bp.id:bp.version \              # buildpack from builder
+  --buildpack file:///home/user/path/to/buildpack/ \          # absolute via file
+  --buildpack /home/user/path/to/buildpack/ \                 # absolute via schemeless
+  --buildpack ../path/to/buildpack/ \                         # relative file
+  --buildpack docker:/cnbs/some-package \                     # Docker Hub image
+  --buildpack docker://gcr.io/cnbs/sample-package-2:bionic \  # GCR image (with tag)
+  --buildpack cnb://index.buildpack.io/my-org/bp.id \         # buildpack from registry
+  --buildpack urn:cnb:builder                                 # All buildpacks in builder
+```   
+
+### Config Files
+
+Config files would replace `image` with `uri`, using the `docker://` scheme, where applicable.
+
+`package.toml`:
+```toml
+[buildpack]
+uri = "."
+
+[[dependencies]]
+uri = "docker:/my/image:latest"
+
+# ...
+``` 
+
+`builder.toml`:
+```toml
+# ...
+
+[[buildpacks]]
+uri = "./my/buildpacks/"
+
+[[buildpacks]]
+uri = "docker:/my/image:latest"
+
+# ...
+```
+
+# Additional considerations
+
+#### `pack` User Experience
+
+`pack` could provide additional logic in order for users to have a nicer user experience when providing references
+via CLI. For example, `pack` could have an order of precedence to make a best guess at which type of reference
+they are referring to or provide additional feedback if it's ambiguous.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The user would need to type an additional scheme prefix for any URI which is not a relative reference.
+
+# Alternatives
+[alternatives]: #alternatives
+
+#### Flags and Config Keys
+
+A different flag or config properties to signify the different protocols and would render the need for
+URI moot.
+
+#### Docker scheme
+
+An alternative to `docker` is to use a more generic term such as `oci`. There is an official
+[OCI distribution spec](https://github.com/opencontainers/distribution-spec/blob/master/spec.md) based on Docker HTTP
+API v2 but it's direct compatibility would require additional research. Additionally, `oci` scheme has an overlap with
+[Oracle Cloud Infrastructure Object Storage Service](https://docs.cloud.oracle.com/en-us/iaas/tools/hdfs/2.9.2.1/). 
+
+# Prior Art
+[prior-art]: #prior-art
+
+- Predefined set of URI schemes for OCI images created by [Skopeo](https://github.com/containers/skopeo).
+- [from=builder][from-builder] syntax.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Where would this be documented long-term?
+    - Answer: Would solely live in the docs site.
+- How does `urn:cnb:registry:bp.id:bp.version` translate to a `namespace` + `name` as detailed in [RFC #22](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md)?
```

Comment:
There is an unresolved question on the original RFC:

- [Should we split id into namespace and id?](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md#unresolved-questions)
    - @jkutner, was there any insight there that might inform this unresolved question?

Reply:
As of now, a `namespace/name` is an acceptable form of buildpack ID (like ID is a superset). So think defining the URN as `urn:cnb:registry:id:version` is sufficient for this proposal. In a future proposal we might define more strict rule about ID, but this definition would stay the same. Does that make sense?